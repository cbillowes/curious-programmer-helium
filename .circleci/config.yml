version: 2
jobs:
  build:
    docker:
      - image: circleci/node:chakracore-8-browsers-legacy
    steps:
      - checkout
      - run: echo "Chopping onions without the tears"
      - run: npm run build
  test:
    docker:
      - image: circleci/node:chakracore-8-browsers-legacy
    steps:
      - checkout
      - run: echo "Making soup and drinking beers"
      - run: sudo npm install -g jest-cli
      - run: npm run test
  page_speed:
    docker:
      - image: circleci/node:chakracore-8-browsers-legacy
    steps:
      - checkout
      - run: echo "Making sure it won't take years"
      - run: echo "* Deploy to Netlify"
      - run: echo "* Execute lighthouse page speed tests on build"
workflows:
  version: 2
  integration:
    jobs:
      - build
      - test:
          requires:
            - build
      - page_speed:
          requires:
            - test
