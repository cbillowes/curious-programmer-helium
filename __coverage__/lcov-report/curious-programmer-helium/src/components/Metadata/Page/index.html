<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for curious-programmer-helium/src/components/Metadata/Page</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../../../../prettify.css" />
    <link rel="stylesheet" href="../../../../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../../../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../../../../index.html">All files</a> curious-programmer-helium/src/components/Metadata/Page
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">82.5% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>33/40</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">64.29% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>18/28</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">100% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>10/10</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">82.5% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>33/40</span>
      </div>
    </div>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Metadata.jsx"><a href="Metadata.jsx.html">Metadata.jsx</a></td>
	<td data-value="82.5" class="pic high"><div class="chart"><div class="cover-fill" style="width: 82%;"></div><div class="cover-empty" style="width:18%;"></div></div></td>
	<td data-value="82.5" class="pct high">82.5%</td>
	<td data-value="40" class="abs high">33/40</td>
	<td data-value="64.29" class="pct medium">64.29%</td>
	<td data-value="28" class="abs medium">18/28</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="10" class="abs high">10/10</td>
	<td data-value="82.5" class="pct high">82.5%</td>
	<td data-value="40" class="abs high">33/40</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sat Jan 05 2019 22:01:48 GMT+0200 (South Africa Standard Time)
</div>
</div>
<script src="../../../../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../../../../sorter.js"></script>
</body>
</html>
