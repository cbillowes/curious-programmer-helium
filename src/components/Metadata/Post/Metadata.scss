@import "../../../layout/layout.scss";

.post-metadata {
  p {
    font-style: italic;
    text-align: right;
    font-size: 0.8em;
  }

  .tags {
    text-align: center;
  }
}