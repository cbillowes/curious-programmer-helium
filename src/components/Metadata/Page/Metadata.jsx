import React, { Component } from "react"
import Helmet from "react-helmet"
import { graphql } from "gatsby";
import config from "../../../../data/SiteConfig"
import navigation from "../../../../data/Navigation"
import Path, { relative } from "path";

class Metadata extends Component {
  constructor(props) {
    super(props)
    this.state = {
      slug: (this.props.data) ? this.props.data.slug.edges[0].node.fields.slug : this.props.slug ? this.props.slug : "/"
    }
  }

  getTitle = (title) => {
    if (title && title !== config.siteTitle) {
      return `${title} | ${config.siteTitle}`
    }

    if (!title) {
      const navigationItem = getNavigationItem(this.state.slug)
      return `${navigationItem.name} | ${config.siteTitle}`
    }

    return config.siteTitle
  }

  getDescription = description => {
    if (description) {
      return description
    }
    return config.siteDescription
  }

  getImageUrl = images => {
    if (images) {
      const image = (images.background) ? images.background : (images.post) ? images.post : config.siteLogo
      if (image.startsWith("http")) {
        return image
      }
      if (image.startsWith("/")) {
        return `${config.siteUrl}${image}`
      }
      return config.siteUrl + Path.normalize(`${this.state.slug}/${image}`)
    }
    return `${config.siteUrl}${config.siteLogo}`
  }

  getUrl = slug => {
    if (slug) {
      return `${config.siteUrl}${slug}`
    }
    let navigationItem = getNavigationItem(this.state.slug)
    let pageUrl = config.siteUrl + navigationItem.url
    return pageUrl
  }

  getType = type => {
    return "website"
  }

  render() {
    const { title, description, images, slug, type } = this.props
    const pageTitle = this.getTitle(title)
    const pageDescription = this.getDescription(description)
    const imageUrl = this.getImageUrl(images)
    const pageUrl = this.getUrl(slug ? slug : this.state.slug)
    const pageType = this.getType(type)

    return (
      <Helmet>
        <title>{pageTitle}</title>
        <link rel="canonical" href={config.siteUrl} />
        <link rel="icon" href="/favicon.png" />
        <meta name="description" content={pageDescription} />
        <meta name="image" content={imageUrl} />

        <meta name="twitter:card" content="summary_large_image" />
        <meta name="twitter:site" content={config.userTwitter} />
        <meta name="twitter:creator" content={config.userTwitter} />

        <meta property="fb:app_id" content={config.siteFBAppID} />
        <meta property="og:title" content={pageTitle} />
        <meta property="og:description" content={pageDescription} />
        <meta property="og:image" content={imageUrl} />
        <meta property="og:url" content={pageUrl} />
        <meta property="og:type" content={pageType} />

        <script type="application/ld+json">
          {getSchema(pageTitle, pageDescription, pageUrl, imageUrl)}
        </script>
      </Helmet>
    );
  }
}

function getNavigationItem(slug) {
  let navigationItem = navigation[0]
  navigation.map(item => {
    if (slug.indexOf(item.url) > -1) {
      navigationItem = item
    }
    return navigationItem
  })
  return navigationItem
}

function getSchema(title, description, url, image) {
  const schema = [
    {
      "@context": "http://schema.org",
      "@type": "WebSite",
      url: url,
      name: config.siteTitle,
      alternateName: config.siteTitleAlt
    },
    {
      "@context": "http://schema.org",
      "@type": "BreadcrumbList",
      itemListElement: [
        {
          "@type": "ListItem",
          position: 1,
          item: {
            "@id": url,
            name: title,
            image: image
          }
        }
      ]
    },
    {
      "@context": "http://schema.org",
      "@type": "BlogPosting",
      url: url,
      name: title,
      alternateName: config.siteTitleAlt,
      headline: title,
      image: {
        "@type": "ImageObject",
        url: image
      },
      description: description
    }
  ]
  return JSON.stringify(schema)
}

export default Metadata

export const pageQuery = graphql`
  query MetadataComponent($slug: String!) {
    allMarkdownRemark {
      edges {
        node {
          fields {
            slug
          }
        }
      }
    }
    markdownRemark(frontmatter: { slug: { eq: $slug } }) {
      fields {
        slug
      }
    }
  }
`