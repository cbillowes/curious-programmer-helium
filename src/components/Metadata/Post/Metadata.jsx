import React, { Component } from "react"
import Moment from "react-moment"
import Tags from "../../Tag/Tags"
import "./Metadata.scss"

class Metadata extends Component {
  render() {
    const { tags, date, timeToRead} = this.props

    return (
      <div className="post-metadata">
        <p>
          posted { " " }
          <Moment date={date} fromNow /> { " " }
          on { " " }
          <Moment date={date} format="ddd, D MMMM YYYY" /> { " " }
          (Est. {timeToRead} minute read)
        </p>
        {(tags) ? <Tags tags={tags} /> : <span> </span>}
      </div>
    )
  }
}

export default Metadata