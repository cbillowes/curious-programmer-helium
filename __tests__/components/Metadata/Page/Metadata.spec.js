import React from "react"
import renderer from "react-test-renderer"
import { mount } from "enzyme"
import Helmet from "react-helmet"
import Metadata from "../../../../src/components/Metadata/Page/Metadata"
import config from "../../../../data/SiteConfig"

function getMetadataTagFromHelmet(helmet, metaProperty) {
  for (let i = 0; i < helmet.metaTags.length; i++) {
    if (helmet.metaTags[i].name === metaProperty || helmet.metaTags[i].property == metaProperty) {
      return helmet.metaTags[i]
    }
  }
}

describe("a meta data component", () => {
  const data = {
    "slug": {
      "edges": [{
        "node": {
          "fields": {
            "slug": "/"
          },
        }
      }]
    }
  }
  it("should render", () => {
    const component = <Metadata data={data} />
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe("an un-configured meta data component", () => {
  const slug = "/tags/technical"
  const data = {
    "slug": {
      "edges": [{
        "node": {
          "fields": {
            "slug": slug
          }
        }
      }]
    }
  }

  const component = mount(<Metadata data={data} />, {
    data: data
  })
  const helmet = Helmet.peek()
  const schema = helmet.scriptTags[helmet.scriptTags.length - 1].innerHTML
  const canonical = helmet.linkTags[0]

  it("should generate schema types", () => {
    expect(schema.indexOf(`"@type":"WebSite"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"@type":"BreadcrumbList"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"@type":"BlogPosting"`)).toBeGreaterThan(-1)
  })

  it("should deduce the title using navigation data", () => {
    expect(helmet.title).toEqual(`Tags | ${config.siteTitle}`)
    expect(schema.indexOf(`"alternateName":"${config.siteTitleAlt}"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"headline":"Tags | ${config.siteTitle}"`)).toBeGreaterThan(-1)
  })

  it("should generate the description using the site description", () => {
    const description = config.siteDescription
    expect(getMetadataTagFromHelmet(helmet, "description").content).toEqual(description)
    expect(getMetadataTagFromHelmet(helmet, "og:description").content).toEqual(description)
    expect(schema.indexOf(`"description":"${description}"`)).toBeGreaterThan(-1)
  })

  it("should generate the image pointing to the site logo", () => {
    const image = `${config.siteUrl}${config.siteLogo}`
    expect(getMetadataTagFromHelmet(helmet, "image").content).toEqual(image)
    expect(getMetadataTagFromHelmet(helmet, "og:image").content).toEqual(image)
    expect(schema.indexOf(`"image":"${image}"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"@type":"ImageObject","url":"${image}"`)).toBeGreaterThan(-1)
  })

  it("should generate the url pointing to the site url", () => {
    const url = `${config.siteUrl}${slug}`
    expect(getMetadataTagFromHelmet(helmet, "og:url").content).toEqual(url)
    expect(schema.indexOf(`"url":"${url}"`)).toBeGreaterThan(-1)
    expect(canonical.href).toEqual(config.siteUrl)
  })

  it("should generate a website open graph type", () => {
    expect(getMetadataTagFromHelmet(helmet, "og:type").content).toEqual("website")
  })

  it("should generate twitter meta data", () => {
    expect(getMetadataTagFromHelmet(helmet, "twitter:card").content).toEqual("summary_large_image")
    expect(getMetadataTagFromHelmet(helmet, "twitter:site").content).toEqual(config.userTwitter)
    expect(getMetadataTagFromHelmet(helmet, "twitter:creator").content).toEqual(config.userTwitter)
  })
})

describe("a configured meta data component", () => {
  const title = "Hello Page"
  const description = "Just a little bit of a write up."
  const slug = "/tags/technical"
  const image = "/image.png"
  const type = "article"
  const component = mount(<Metadata
    title={title}
    description={description}
    images={{ post: image }}
    slug={slug}
    type={type} />)
  const helmet = Helmet.peek()
  const schema = helmet.scriptTags[helmet.scriptTags.length - 1].innerHTML

  it("should generate using that title", () => {
    expect(helmet.title).toEqual(`${title} | ${config.siteTitle}`)
    expect(schema.indexOf(`"alternateName":"${config.siteTitleAlt}"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"headline":"${title} | ${config.siteTitle}"`)).toBeGreaterThan(-1)
  })

  it("should generate the description using the site description", () => {
    expect(getMetadataTagFromHelmet(helmet, "description").content).toEqual(description)
    expect(getMetadataTagFromHelmet(helmet, "og:description").content).toEqual(description)
    expect(schema.indexOf(`"description":"${description}"`)).toBeGreaterThan(-1)
  })

  it("should generate the image url", () => {
    const imagePath = `${config.siteUrl}${image}`
    expect(getMetadataTagFromHelmet(helmet, "image").content).toEqual(imagePath)
    expect(getMetadataTagFromHelmet(helmet, "og:image").content).toEqual(imagePath)
    expect(schema.indexOf(`"image":"${imagePath}"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"@type":"ImageObject","url":"${imagePath}"`)).toBeGreaterThan(-1)
  })

  it("should generate the url pointing to the site url", () => {
    const url = `${config.siteUrl}${slug}`
    expect(getMetadataTagFromHelmet(helmet, "og:url").content).toEqual(url)
    expect(schema.indexOf(`"url":"${url}"`)).toBeGreaterThan(-1)
  })

  it("should generate a website open graph type", () => {
    expect(getMetadataTagFromHelmet(helmet, "og:type").content).toEqual("website")
  })
})

describe("the images configured for a meta data component", () => {
  it("should generate the url based on the background image", () => {
    const background = "/background-photo.png"
    const component = mount(<Metadata images={{ background: background }} />)
    const helmet = Helmet.peek()
    const schema = helmet.scriptTags[helmet.scriptTags.length - 1].innerHTML

    const imagePath = `${config.siteUrl}${background}`
    expect(getMetadataTagFromHelmet(helmet, "image").content).toEqual(imagePath)
    expect(getMetadataTagFromHelmet(helmet, "og:image").content).toEqual(imagePath)
    expect(schema.indexOf(`"image":"${imagePath}"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"@type":"ImageObject","url":"${imagePath}"`)).toBeGreaterThan(-1)
  })

  it("should generate the url based on the post image", () => {
    const post = "/post.png"
    const component = mount(<Metadata images={{ post: post }} />)
    const helmet = Helmet.peek()
    const schema = helmet.scriptTags[helmet.scriptTags.length - 1].innerHTML

    const imagePath = `${config.siteUrl}${post}`
    expect(getMetadataTagFromHelmet(helmet, "image").content).toEqual(imagePath)
    expect(getMetadataTagFromHelmet(helmet, "og:image").content).toEqual(imagePath)
    expect(schema.indexOf(`"image":"${imagePath}"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"@type":"ImageObject","url":"${imagePath}"`)).toBeGreaterThan(-1)
  })

  it("should generate the site logo when there is an image not catered for", () => {
    const image = "/image.png"
    const component = mount(<Metadata images={{ image: image }} />)
    const helmet = Helmet.peek()
    const schema = helmet.scriptTags[helmet.scriptTags.length - 1].innerHTML

    const imagePath = `${config.siteUrl}${config.siteLogo}`
    expect(getMetadataTagFromHelmet(helmet, "image").content).toEqual(imagePath)
    expect(getMetadataTagFromHelmet(helmet, "og:image").content).toEqual(imagePath)
    expect(schema.indexOf(`"image":"${imagePath}"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"@type":"ImageObject","url":"${imagePath}"`)).toBeGreaterThan(-1)
  })

  it("should return the absolute image url", () => {
    const background = "https://picsum.photo/#/64545"
    const component = mount(<Metadata images={{ background: background }} />)
    const helmet = Helmet.peek()
    const schema = helmet.scriptTags[helmet.scriptTags.length - 1].innerHTML

    expect(getMetadataTagFromHelmet(helmet, "image").content).toEqual(background)
    expect(getMetadataTagFromHelmet(helmet, "og:image").content).toEqual(background)
    expect(schema.indexOf(`"image":"${background}"`)).toBeGreaterThan(-1)
    expect(schema.indexOf(`"@type":"ImageObject","url":"${background}"`)).toBeGreaterThan(-1)
  })
})

describe("the image url relative to a slug generated for a meta data component", () => {
  it("should use its relative path", () => {
    const slug = "/tags/technical"
    const background = "background-photo.png"
    const component = mount(<Metadata slug={slug} images={{ background: background }} />)
    const helmet = Helmet.peek()
    expect(getMetadataTagFromHelmet(helmet, "image").content).toEqual(`${config.siteUrl}${slug}/${background}`)
  })

  it("should use its relative path", () => {
    const slug = "/one/two/three/four/technical"
    const background = "../../../background-photo.png"
    const component = mount(<Metadata slug={slug} images={{ background: background }} />)
    const helmet = Helmet.peek()
    expect(getMetadataTagFromHelmet(helmet, "image").content).toEqual(`${config.siteUrl}/one/two/background-photo.png`)
  })

  it("should use its relative path", () => {
    const slug = "/one/two/three/four/technical"
    const background = "/background-photo.png"
    const component = mount(<Metadata slug={slug} images={{ background: background }} />)
    const helmet = Helmet.peek()
    expect(getMetadataTagFromHelmet(helmet, "image").content).toEqual(`${config.siteUrl}/background-photo.png`)
  })
})

describe("the page url generated for a meta data component", () => {
  it("should generate the site url", () => {
    const component = mount(<Metadata />)
    const helmet = Helmet.peek()
    expect(getMetadataTagFromHelmet(helmet, "og:url").content).toEqual(`${config.siteUrl}/`)
  })

  it("should generate the site url for that page", () => {
  const slug = "/tags/technical"
  const data = {
    "slug": {
      "edges": [{
        "node": {
          "fields": {
            "slug": slug
          },
        }
      }]
    }
  }
    const component = mount(<Metadata data={data} />)
    const helmet = Helmet.peek()
    expect(getMetadataTagFromHelmet(helmet, "og:url").content).toEqual(`${config.siteUrl}${slug}`)
  })
})