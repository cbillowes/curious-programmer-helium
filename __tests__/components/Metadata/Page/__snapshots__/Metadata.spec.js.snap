// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a meta data component should render 1`] = `null`;
